### PR TITLE
feat: Add initial support for attachments

### PR DESCRIPTION
**Pull request by dcramer**
Add support for storing event attachments using the existing filestorage service and a new EventAttachment model.

### CODE REVIEW

**Comment by dcramer** [comment]
Putting this up for schema review per conversation about supporting attachments on events.

We wont yet allow submission of these as we need to implement a number of things yet, including:

- metrics around attachment volume
- user feedback-type endpoint for uploading and processing attachments
- tbd mechanism to support attachments as part of minidump processing

File: src/sentry/models/eventattachment.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+from django.db import models
+from django.utils import timezone
+
+from sentry.db.models import BoundedBigIntegerField, FlexibleForeignKey, Model, sane_repr
+
+
+class EventAttachment(Model):
+    __core__ = False
+
+    project_id = BoundedBigIntegerField()
```
Comment (dcramer) [inline src/sentry/models/eventattachment.py]: this model follows the same principles as UserReport where we expect to be able to handle an attachment uploaded before a Group is associated

**Review by dcramer — COMMENTED** [review]
I need to add a deletion handler for this yet

**Comment by dcramer** [comment]
Ive fixed a few deletion bugs with Events primarily. Added tests for them as well.

**Comment by mitsuhiko** [comment]
Is there a specific reason you envision user uploads of event attachments?

Generally I think this approach is fine as we're already considering two cases most likely: associated logfile attachments and minidumps.

**Comment by dcramer** [comment]
@mitsuhiko this is per the conversation we already had about requirements for minidumps, and yes its associated with that kind of data

you can also imagine someone uploading video, screenshots, etc

**Comment by mitsuhiko** [comment]
@dcramer i'm asking because minidump attachments will happen as part of event submission, same with log file data from the mindump submission. I do not currently see a usecase for externally attaching things from the client SDK.

**Comment by dcramer** [comment]
@mitsuhiko its not guaranteed they'd be uploaded in the same request, so while in that one specific case they might, that doesnt mean its the only way they'd possible be sent (again, think of it like user feedback, that it might make sense to collect attachments after the reports already generated and sent to sentry, or as part of an async process)

**Comment by mitsuhiko** [comment]
@dcramer are we still talking minidumps here or something else? In the minidump case they are sent in the same request.

File: src/sentry/models/eventattachment.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    group_id = BoundedBigIntegerField(null=True)
+    event_id = models.CharField(max_length=32)
+    file = FlexibleForeignKey('sentry.File')
+    name = models.TextField()
+    date_added = models.DateTimeField(default=timezone.now)
```
Comment (mattrobenolt) [inline src/sentry/models/eventattachment.py]: We can add this later, but probably useful to have an index on date_added if we did want to TTL?

**Review by mattrobenolt — APPROVED** [review]

File: src/sentry/models/eventattachment.py (hunk 2 above)
Comment (dcramer) [reply src/sentry/models/eventattachment.py]: @mattrobenolt (project_id, date_added)?

Comment (mattrobenolt) [reply src/sentry/models/eventattachment.py]: Yeah, if you'd want to have potentially different TTLs per project or something.

We can always add this later, so I don't think it's a big deal now to hypothesize what a use case might be.

**Comment by dcramer** [comment]
Updated this per a recent conversation that we should probably start restricting things better with feature flag.

**Comment by dcramer** [comment]
Patch should be good now:

- added (project_id, date_added)
- expanded deletion to Project cascades

Also to finish up conversation: this is only scaffolding. I intend to expose basic UI for downloading attachments (behind the feature), and metrics, but wont be building any way of actually uploading the attachments.

We'll likely need to support at least two ways of processing events:

1. Multi-part uploads, as this seems to be what crashpad is aiming for natively
2. Future: a mechanism for non-crashpad uploads (this could be the same approach, multi-part, but will likely depend on SDKs and runtimes)

**Pull request merged**